design-book/src/roadmap.md

```diff
@@ -0,0 +1,100 @@
+For actionable, well-scoped items, please open issues instead.
+
+As you complete tasks, check them off here!
+The tasks don't need to be done *well* to be considered completed here: the Bevy community is very good at cleaning up loose strings as we work.
+
+Feel free to tackle any listed item in any order: foundations and primitives can often be built out well before the features that need them are in place.
+This list is not "all of the features that the editor will have": other stages will follow as these features are added.
+It's best not to plan too far in advance!
+
+## Stage 0: Hello Project
+
+- [x] make a repo
+- [x] add a README
+- [x] add a CONTRIBUTING.md
+- [x] add a basic Bevy project
+- [x] define the broad plan for what we're doing
+- [x] define basic constraints for how we're building the editor
+- [x] define a shared vision for what the editor is (and is not)
+
+## Stage 1: Technically an Editor
+
+- [ ] can load scenes from disk using a native file picking widget
+- [ ] can display scenes in a viewer
+  - [ ] 2D
+  - [ ] 3D
+- [ ] simple camera controller
+  - [ ] 2D
+  - [ ] 3D
+- [ ] gizmos can be toggled
+  - [ ] UI outlines
+  - [ ] lights
+  - [ ] AABBs
+- [ ] the list of entities and their components is shown in the inspector
+  - [ ] supports hierarchy via a folding tree view
+- [ ] existing component values can be modified using a graphical interface
+- [ ] resources can be inspected, showing their values and allowing modification
+- [ ] loaded assets can be inspected, providing basic information about them
+- [ ] scenes can be saved back to disk
+
+## Stage 2: Editor-Game Interaction
+
+- [ ] users can press a button in the editor, launching the game from the very beginning
+- [ ] users can press a button in the editor, and their game will run, loading the currently active scene
+- [ ] components can be added and removed, including components specific to the user's game
+- [ ] entities can be spawned and despawned via a GUI
+- [ ] entity hierarchies can be spawned and despawned via a GUI
+  - [ ] manually
+  - [ ] by composing scene files
+- [ ] game-specific rendering techniques display correctly in the editor's scene editing
+- [ ] distinct editor and game views that can be run simultaneously
+
+## Stage 3: UI/UX Foundations
+
+- [ ] tooltips
+- [ ] hotkeys
+- [ ] primitive widgets
+  - [ ] text entry
+  - [ ] numeric entry
+  - [ ] radio buttons
+  - [ ] dropdown
+  - [ ] checkboxes
+- [ ] editor UI foundations
+  - [ ] resizable panes
+  - [ ] scrollable tool detail panels
+  - [ ] toolbar
+  - [ ] Blender-style workspaces
+  - [ ] context menus
+  - [ ] command palette
+- [ ] dedicated hotkeys and controls for the most common operations:
+  - [ ] manipulating camera
+  - [ ] modifying transforms
+  - [ ] toggling visibility
+  - [ ] adding and removing entities
+  - [ ] saving the project
+  - [ ] opening the command palette
+
+## Stage 4: Fundamental Features
+
+- [ ] undo-redo abstraction
```

Comment:
I would suggest adding this earlier than step 4. Since I've had PR hanging for undo/redo since January 👀 https://github.com/bevyengine/bevy_editor_prototypes/pull/2

Reply:
By 'undo/redo' what we really mean is 'define our core datamodel and it's api'. I also don't think this should be pushed till stage 4.

Reply:
> By 'undo/redo' what we really mean is 'define our core datamodel and it's api'. If you don't design your state mutations to be atomic, it will not be easy to retrofit. I also don't think this should be pushed till stage 4.

My proposed undo/redo model is fairly easily expandable and atomic. And I'd say my PR would be a good starting point for a better api in the future. And yes, undo/redo is very fundamental and late implementation can force a lot of code rewriting.

Reply:
Giving it a review.